src/vsr/journal.zig

```diff
@@ -1356,93 +1355,141 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
+
+                    if ((op_max == op_prepare_max - 1 and index == op_prepare_max_slot.index) or
+                        op_max_to_op_prepare_max.contains(slot))
+                    {
+                        const header_prepare_untrusted = &journal.headers[index];
+                        const header_prepare_ok = header_ok(
+                            replica.cluster,
+                            slot,
+                            header_prepare_untrusted,
+                        );
+                        const header_redundant_ok = header_ok(
+                            replica.cluster,
+                            slot,
+                            &journal.headers_redundant[index],
+                        );
+
+                        // We need our head op to be certain to reliably truncate torn prepares.
+                        // Head op is uncertain if we encounter one of the below faults:
+
+                        // 1. Corrupt redundant header or a misdirected read to a redundant header.
+                        if (header_redundant_ok == null) return torn_slots[0..0];
+
+                        // 2. Redundant header is set to .reserved. Could happen if:
+                        //    i. Slot was found corrupt on a previous startup, which set the header
+                        //       to reserved in memory.
+                        //    ii. Replica crashes *before* the corrupt slot was repaired, but
+                        //       *after* the reserved header was written to disk with a write to
+                        //       a nearby header (there are multiple headers in a single sector)
+                        if (header_redundant_ok.?.operation == .reserved) return torn_slots[0..0];
+
+                        // 3. Misdirected read for a prepare header.
+                        if (header_prepare_untrusted.valid_checksum()) {
+                            // This is a faulty slot and header is valid, prepare must be corrupt.
+                            assert(header_prepare_ok == null);
+                            return torn_slots[0..0];
+                        }
+
+                        // Header is valid and from a previous wrap.
+                        assert(header_redundant_ok != null);
+                        assert(header_redundant_ok.?.op < op_max);
+
+                        assert(!header_prepare_untrusted.valid_checksum());
+                        assert(!journal.prepare_inhabited[index]);
+
+                        torn_slots[torn_slot_count] = slot;
+                        torn_slot_count += 1;
                     }
                 }
             }
-
-            // The prepare is torn.
-            assert(!journal.prepare_inhabited[torn_slot.index]);
-            assert(!torn_prepare_untrusted.valid_checksum());
-            assert(cases[torn_slot.index].decision(replica.solo()) == .vsr);
-            return torn_slot;
+            return torn_slots[0..torn_slot_count];
```

Comment:
Since this array is on our stack, I think returning a slice of it is UB.
Maybe use a `BoundedArray` instead?

Reply:
Oops, this was a bad miss on my part! Thank you for pointing it out.

Reply:
Took care of this in https://github.com/tigerbeetle/tigerbeetle/pull/2345/commits/8c0600bcefe1bab79a1df21d142fbe2c3d4a10d3.